doc/design/clustered-etcd-alert-store/design.md

```diff
@@ -0,0 +1,150 @@
+# Clustered Etcd Alert Store
+
+Type: Design Document (and PR)
+
+Date: 2019-09-20
+
+Author: David C Wang <dcwangmit01@gmail.com>
+
+## Status Quo
+
+The current Alertmanager clustering implementation will allow multiple
+instances of Alertmanager to sync recently-updated notification log entries and
+alert silences via the gossip protocol.  The syncing of alert state is not yet
+implemented.
+
+Currently, Alertmanager alert state is entirely stored in process memory.  It
+is lost upon restart of Alertmanager.  This, in conjunction with the fact that
+Alertmanager alert state is not synced with clustered peers presents several
+problems:
+
+* Alert emitters must send alerts to every single Alertmanager instance in the
+  cluster.  If some Alertmanager instances receive alerts that other instances
+  do not, their state will drift apart.
+
+  * Typically, the alert emitter is Prometheus which uses service discovery to
+    locate all Alertmanager instances, and then sends alerts to each individual
+    target.  Custom alert emitters may not have that service discovery
+    capability and must be explicity configured.
+
+* Because the alert must be sent to every Alertmanager instance, this precludes
```

Comment:
This design doesn't change that, for reliability you still need to send it to multiple Alertmanagers. Comparing with stateless services isn't really useful.

Reply:
The effect of this PR is that the alert does go to each Alertmanager in the cluster, even if it is sent to a single Alertmanager instance.  If an alert gets to any individual Alertmanager instance, the alert state, if it is a new fingerprint or has been modified, is written to Etcd.  Then, the other Alertmanagers in the cluster will receive the updated alert state through the *etcd watch* clients that they have initiated.